src/main/java/com/google/api/codegen/Resources.java

```diff
@@ -69,22 +72,48 @@
     // (useful for testability).
     Map<String, FieldSegment> specs = new TreeMap<>();
 
+    Map<String, List<FieldSegment>> methodFieldSegs = getMethodFieldSegmentMapFromHttpPaths(methods);
+    for (List<FieldSegment> segList : methodFieldSegs.values()) {
+      for (FieldSegment fieldSegment : segList) {
+        String resourcePath = PathSegment.toSyntax(fieldSegment.getSubPath());
+        // If there are multiple field segments with the same resource path, the last
+        // one will be used, making the output deterministic. Also, the first field path
+        // encountered tends to be simply "name" because it is the corresponding create
+        // API method for the type.
+        specs.put(resourcePath, fieldSegment);
+      }
+    }
+    return specs.values();
+  }
+
+  /**
+   * Returns a map from method name to a list of FieldSegment objects. Each FieldSegment will be
+   * composed of a field name and a path segment, representing an abstract resource name with
+   * wildcards.
+   */
+  public static Map<String, List<FieldSegment>> getMethodFieldSegmentMapFromHttpPaths(
```

Comment:
This whole class could be moved under com.google.api.codegen.config; the only reference to it outside of *.codegen.config is the isIdempotent method, but I think that can just be moved into GapicContext, the only place that calls it.